pow/src/compute/mod.rs

```diff
@@ -89,34 +92,82 @@ fn need_new_vm<M: randomx::WithCacheMode>(
 	need_new_vm
 }
 
+fn do_new_vm<M: randomx::WithCacheMode>(
+	key_hash: &H256,
+	machine: &RefCell<Option<(H256, randomx::VM<M>)>>,
+	shared_caches: &Mutex<LruCache<H256, Arc<randomx::Cache<M>>>>,
+	f_has_large_pages: fn(&Config) -> bool,
+) -> Result<(), ()> {
+	let mut shared_caches = shared_caches.lock();
+
+	if !f_has_large_pages(global_config()) {
+		if let Some(cache) = shared_caches.get_mut(key_hash) {
+			machine.replace(Some((*key_hash, randomx::VM::new(cache.clone(), global_config()))));
+
+			Ok(())
+		} else {
+			info!(
+				target: "kulupu-randomx",
+				"At block boundary, generating new RandomX {} cache with key hash {} ...",
+				M::description(),
+				key_hash,
+			);
+
+			let cache = match randomx::Cache::new(&key_hash[..], global_config()) {
+				Ok(cache) => Arc::new(cache),
+				Err(_) => panic!("Cache allocation failed"),
+				};
+
+			shared_caches.insert(*key_hash, cache.clone());
+			machine.replace(Some((*key_hash, randomx::VM::new(cache, global_config()))));
+
+			Ok(())
+		}
+	} else {	
+		if let Some(cache) = shared_caches.get_mut(key_hash) {
+			machine.replace(Some((*key_hash, randomx::VM::new(cache.clone(), global_config()))));
+
+			Ok(())
+		} else {
+			machine.replace(None);
+			let key_to_replace = match (*shared_caches)
+					.iter()
+					.find(|&(_, cache)| Arc::strong_count(cache) == 1)
+					.and_then(|(key, _)| Some(*key)) {
+				Some(key) => key,
+				None => return Err(()),
+			};
+			let mut cache = shared_caches
+					.remove(&key_to_replace)
+					.unwrap();
+			Arc::get_mut(&mut cache)
+					.unwrap()
+					.reinit(&key_hash[..]);
+			shared_caches.insert(*key_hash, cache.clone());
+			machine.replace(Some((*key_hash, randomx::VM::new(cache, global_config()))));
+
+			Ok(())
+		}
+	}
+}
+
 fn loop_raw_with_cache<M: randomx::WithCacheMode, FPre, I, FValidate, R>(
 	key_hash: &H256,
 	machine: &RefCell<Option<(H256, randomx::VM<M>)>>,
 	shared_caches: &Arc<Mutex<LruCache<H256, Arc<randomx::Cache<M>>>>>,
 	mut f_pre: FPre,
 	f_validate: FValidate,
+	f_has_large_pages: fn(&Config) -> bool,
 	round: usize,
 ) -> Option<R> where
 	FPre: FnMut() -> (Vec<u8>, I),
 	FValidate: Fn(H256, I) -> Loop<Option<R>>,
 {
 	if need_new_vm(key_hash, machine) {
-		let mut ms = machine.borrow_mut();
-
-		let mut shared_caches = shared_caches.lock().expect("Mutex poisioned");
+		if do_new_vm(key_hash, machine, shared_caches, f_has_large_pages).is_err() {
+			thread::sleep(Duration::new(1, 0));
 
-		if let Some(cache) = shared_caches.get_mut(key_hash) {
-			*ms = Some((*key_hash, randomx::VM::new(cache.clone(), global_config())));
-		} else {
-			info!(
-				target: "kulupu-randomx",
-				"At block boundary, generating new RandomX {} cache with key hash {} ...",
-				M::description(),
-				key_hash,
-			);
-			let cache = Arc::new(randomx::Cache::new(&key_hash[..], global_config()));
-			shared_caches.insert(*key_hash, cache.clone());
-			*ms = Some((*key_hash, randomx::VM::new(cache, global_config())));
+			return None
```

Comment:
This line will not work. It would break any function (including `compute` below) which expects `f_validate` to follow some constraints for its return value. You need to change the return value to `Result<Option<R>, String>` and then handle errors that way.

Reply:
I used an enum for errors, i guess it's fine